solutions/silver/usaco-895.mdx

```diff
@@ -2,234 +2,272 @@
-ll area = 0;
-ll perimeter = 0;
-
-void setIO(string name) {
-	ios_base::sync_with_stdio(0);
-	cin.tie(0);
-	if (len(name)) {
-		freopen((name + ".in").c_str(), "r", stdin);
-		freopen((name + ".out").c_str(), "w", stdout);
-	}
-}
+			perimeter += sides;  // Add remaining sides to the perimeter
+		}
 
-void floodfill(int r, int c) {
-	if (r < 0 || r >= n || c < 0 || c >= n) {
-		currP++;
-		return;  // if outside grid
-	}
-	if (visited[r][c]) return;  // already visited this square
-	if (grid[r][c] == '.') {
-		currP++;
-		return;
+		return new int[] {area, perimeter};
 	}
-	visited[r][c] = true;  // mark current square as visited
-	if (grid[r][c] == '#') { currA++; }
-	// recursively call floodfill for neighboring squares
-	floodfill(r, c + 1);
-	floodfill(r, c - 1);
-	floodfill(r - 1, c);
-	floodfill(r + 1, c);
-}
 
-int main() {
-	setIO("perimeter");
-	cin >> n;
-	for (int i = 0; i < n; i++) {
-		for (int j = 0; j < n; j++) { cin >> grid[i][j]; }
-	}
-	for (int i = 0; i < n; i++) {
-		for (int j = 0; j < n; j++) {
-			if (!visited[i][j] && grid[i][j] == '#') {
-				floodfill(i, j);
-				if (area == currA) {
-					perimeter = min(perimeter, currP);
-				} else if (currA > area) {
-					area = currA;
-					perimeter = currP;
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new FileReader("perimeter.in"));
+		int n = Integer.parseInt(br.readLine());
+
+		char[][] grid = new char[n][n];
+		boolean[][] visited = new boolean[n][n];
+
+		for (int i = 0; i < n; i++) {
+			String line = br.readLine();
+			grid[i] = line.toCharArray();
+		}
+
+		int maxArea = 0;
+		int minPerimeter = Integer.MAX_VALUE;
+		for (int i = 0; i < n; i++) {
+			for (int j = 0; j < n; j++) {
+				if (grid[i][j] != '#' || visited[i][j]) { continue; }
+
+				int[] result = bfs(i, j, grid, visited, n);
+				int area = result[0];
+				int perimeter = result[1];
+
+				// Update maxArea and minPerimeter based on the BFS results
+				if (area > maxArea || (area == maxArea && perimeter < minPerimeter)) {
+					maxArea = area;
+					minPerimeter = perimeter;
 				}
-				currA = currP = 0;
 			}
 		}
+
+		PrintWriter pw = new PrintWriter("perimeter.out");
+		pw.println(maxArea + " " + minPerimeter);
+		pw.close();
 	}
-	cout << area << " " << perimeter << endl;
-	return 0;
 }
 ```
 
-</CPPSection>
-
+</JavaSection>
 <PySection>
+
 ```py
-# This solution does TLE on the USACO website on one test case because python is slow.
 from collections import deque
+from typing import List, Tuple
 
-with open("perimeter.in") as r:
-	t = r.readline
-	n = int(t())
-	ice = []
-	visited = [[False] * n for _ in range(n)]
+# Movement directions (up, down, left, right)
+DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]
 
-	for _ in range(n):
-		ice.append(list(t()))
 
-max_area = 0
-min_peri = float("inf")
+def bfs(
+	row: int, col: int, grid: List[List[str]], visited: List[List[bool]], n: int
```

Comment:
you can probably just read in `grid` as a list of strings since you can index them asw